0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,264 @@
+}
+
+// Metadata associated with an object.
+type ObjectMetadata map[string]string
+
+// Error constants:
+var ErrNotFound = errors.New("...")
+var ErrConflict = errors.New("...")
+```
+
+The `key` parameter is the name of the object. Unlike with state stores, there are no key prefixes when operating with objects: applications have full access to all data in the storage service.
+
+Each object can have metadata associated to it, which is normally passed to clients as headers (more on that below). In particular, `Content-Type` is yet another metadata key, treated no differently from anything else.
+
+> Option: we could support auto-detection of the content type if users pass `"auto"` as value, using a library such as [ItalyPaleAle/file-type-stream-go](https://github.com/ItalyPaleAle/file-type-stream-go) (which I'd be very happy to transfer to the Dapr org).
+
+The only exception is the `tag`, which is passed separately from the metadata object. The reason is that in the `SetObject` method, the tag is both an (optional) parameter and a return value. Note however that support for tags / ETags, is not a requirement if the underlying storage service doesn't support it.
+
+When invoking `GetObject` or `SetObject`, the output and input streams (respectively) are created outside of the state store and are owned by the caller. The `GetObject` and `SetObject` methods synchronously return only after all data has been written or read (respectively), and after that the caller can call `Close()` on the streams (if necessary/appropriate).
+
+Notes:
+
+- This building block will be stream-first. It will be working with input and output data as binary streams, and will not perform any transformation or encoding on the data. Thanks to using streams, there are no limits on the size of the input/output data, bypassing `MaxBodySize`.
+- Dapr only supports the "last-write-wins" concurrency pattern for object storage (unlike with state stores). This makes sense considered the intended usage of object storage, which is storing documents that usually do not change. When documents do change, using last-write-wins is consistent with how regular filesystems work, for example.
+- Dapr will not calculate and store the checksum of objects. This is because checksums must be transmitted as header, but Dapr stores data in the state store in a streaming way, so it can't compute the full checksum until the end of the stream. Apps that want to store checksums (such as the `Content-MD5` header) should compute it beforehand and submit it as metadata header.
+
+### HTTP APIs
+
+Applications that interact with object storage using HTTP can leverage these APIs, that follow a more REST-like pattern than the typical state store APIs.
+
+> These APIs have been modeled after the APIs supported by AWS S3 and Azure Blob Storage.
+
+#### Retrieving an object
+
+To retrieve an object, clients make a GET request:
+
+```text
+GET http://localhost:<port>/v1.0-alpha1/objects/<component>/<key>
```

Comment:
So, we would call it the "objects" API and not "blob"? I think "blob" is clear about the unstructured nature of it while "object" reminds me of JSON. It might just be me but food for thought.

Reply:
I had the same doubt. I landed with "object" because the industry-standard term seems to be "object storage", but I agree that "blob" does sound better. I'm fine either way.

Reply:
It would be great for people to vote on the naming for this here.

Reply:
I think that the points you 2 just mentioned here should be part of the overview: about both _blob_ and _objects_ being commonly used to refer to the concept at hand and that the industry usually uses "object".

I will say that _blob_ makes it non-ambiguous to me to what you are targeting with this building block.

Reply:
Arguably "Object storage" contains more than just a binary blob, there's often metadata attached to it -- [AWS](https://aws.amazon.com/what-is/object-storage/) and [Google](https://cloud.google.com/learn/what-is-object-storage) whereas blob storage usually refers to an opaque large binary object  so I think naming matters; are we purely offering the opaque data or something that requires metadata be associated with it?

Reply:
Metadata _can_ be added (as described in the doc) but isn't required.

Reply:
I'd go for object storage here since metadata while optional can play a major part in how the API behaves, and this supports the same object storage notion for AWS and GCP which has metadata as an optional construct.

Reply:
Then I would agree that it feels more like an object store than a blob store, which is fine because objects can be blobs (i.e no metadata associated), but blobs can't really be objects by their accepted definitions.